lonboard/experimental/traits.py

```diff
@@ -95,9 +94,104 @@ def validate(
                         "matplotlib.colors.to_rgba."
                     ),
                 )
-                return
 
             return tuple(map(int, (np.array(c) * 255).astype(np.uint8)))
 
         self.error(obj, value)
         assert False
+
+
+class NormalAccessor(FixedErrorTraitType):
+    """
+    A representation of a deck.gl normal accessor
+
+    Acceptable inputs:
+    - A numpy ndarray with two dimensions: The size of the
+    second dimension must be 3 i.e. `(N,3)`
+    - a pyarrow `FixedSizeListArray` or
+    `ChunkedArray` containing `FixedSizeListArray`s
+    where the size of the inner fixed size list 3.
+    """
+
+    default_value = [0, 0, 1]
+    info_text = (
+        "List representing normal of each object, in [nx, ny, nz]. or numpy ndarray or "
+        "pyarrow FixedSizeList representing the normal of each object, in [nx, ny, nz]"
+    )
+
+    def __init__(
+        self: TraitType,
+        *args,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.tag(sync=True, **NORMAL_SERIALIZATION)
+
+    def validate(
+        self, obj, value
+    ) -> Union[Tuple[int, ...], List[int], pa.ChunkedArray, pa.FixedSizeListArray]:
+        """
+        Values in acceptable types must be contiguous
+        (the same length for all values) and of floating point type
+        """
+
+        fixed_list_size = 3
+
+        if isinstance(value, List):
+            if len(value) != 3:
+                self.error("Normal array must have length 3, (x,y,z)")
+
+            if not all(isinstance(item, float) for item in value):
+                self.error("All elements of Normal array must be floating point type")
+
+            return pa.FixedSizeListArray.from_arrays(value, fixed_list_size)
+
+        if isinstance(value, np.ndarray):
+            if value.ndim != 2 or value.shape[1] != 3:
+                self.error(obj, value, info="Normal array must be 2D with shape (N,3)")
+
+            if not np.any(value == 1):
+                self.error(
+                    obj, value, info="One of the normal vector elements must be 1."
+                )
```

Comment:
This is a great idea for a validation step, but I don't _think_ it's correct. There's no good documentation in deck.gl of the exact format expected. Most likely, these normal vectors are expected to be "normalized" which means that their _length_ is expected to be 1. I.e. x^2 + y^2 + z^2 = 1. This means that almost never will one of the elements themselves be 1.

In any case I'd tend to think this is something we shouldn't validate, because deck.gl might be lenient and allow non-normalized vectors.